src/network/networkd-sriov.c

```diff
@@ -250,28 +249,15 @@ static int link_set_sr_iov_virt_ports(Link *link) {
         if (r < 0)
                 return r;
 
-        r = sd_device_get_syspath(pci_dev, &pci_syspath);
-        if (r < 0)
-                return r;
-
-        dir = opendir(pci_syspath);
-        if (!dir)
-                return -errno;
-
-        FOREACH_DIRENT_ALL(de, dir, break) {
-                _cleanup_(sd_device_unrefp) sd_device *pci_virtfn_dev = NULL;
-
-                if (de->d_type != DT_LNK)
-                        continue;
+        FOREACH_DEVICE_CHILD_WITH_NAME(pci_dev, child, name) {
+                const char *n;
 
                 /* Accept name prefixed with "virtfn", but refuse "virtfn" itself. */
-                if (isempty(startswith(de->d_name, "virtfn")))
-                        continue;
-
-                if (sd_device_new_child(&pci_virtfn_dev, pci_dev, de->d_name) < 0)
+                n = startswith(name, "virtfn");
```

Comment:
maybe add a helper `const char *device_name_startswith(sd_device *d, const char *prefix)`?

Internally that would just do startswith(sd_device_get_sysname(…), …)

Reply:
As commented in the above, virtfn0 is not a sysname...

Another possible option is making `sd_device_get_child_first()` and `_next()` also take a PCRE filter, like we already accept in `sd_device_enumerator_add_match_property()`.
But, that sounds overkill for me.